### PR TITLE
SPDK LEP

### PR DESCRIPTION
**Pull request by DamiaSan**
A new proposal for reimplementing Longhorn Engine with SPDK

### CODE REVIEW

**Comment by liyimeng** [comment]
Hope something do happen this time :D

File: enhancements/20221213-reimplement-longhorn-engine-with-SPDK.md [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+* Implement all actual `longhorn-engine` functionalities
+* Continue to support multiple `longhorn-engine` versions concurrently
+* Maintain as much as possible the same user experience between Longhorn with and without SPDK
+* Lay the groundwork for extending Longhorn to sharding and aggegration of storage devices
+
+
+## Proposal
+
+Create an external orchestrator that, with JSON-RPC calls towards multiple instances of `spdk_tgt` app running in different machines, could manage the durability and reliability of data. Actually, not all needed functionalities to do that are already available in SPDK, so some new JSON-RPC commands will be developed over SPDK. The name of such an orchestrator could be `longhorn-spdk-agent`, taking as example [Storage Management Agent](https://spdk.io/doc/sma.html) that has some similar functionalities.
```
Comment (derekbit) [inline enhancements/20221213-reimplement-longhorn-engine-with-SPDK.md:37]: Can you briefly introduce what the JSON-RPC commands will be used in Longhorn?

Comment (derekbit) [reply enhancements/20221213-reimplement-longhorn-engine-with-SPDK.md:37]: The `IP RPC` is `TCP/IP RPC`, right?

File: enhancements/20221213-reimplement-longhorn-engine-with-SPDK.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+* [Logical volumes](https://spdk.io/doc/logical_volumes.html) library is a flexible storage space management system. It allows creating and managing virtual block devices with variable size on top of other bdevs. The SPDK Logical Volume library is built on top [Blobstore](https://spdk.io/doc/blob.html) which is a persistent, power-fail safe block allocator designed to be used as the local storage system backing a higher level storage service, typically in lieu of a traditional filesystem. Logical volumes have a couple of features like Thinly Provisioning and Snapshots similar to what actual Longhorn-Engine provides.
+* [NVMe over Fabrics](https://spdk.io/doc/nvmf.html) is a feature to presents block devices over a fabrics such as Ethernet, supporting RDMA and TCP transports. The standard Linux kernel initiators for NVMe-oF interoperate with these SPDK NVMe-oF targets, so with this feature we can serve bdev over the network or to other processes
+
+
+## Motivation
```
Comment (derekbit) [inline enhancements/20221213-reimplement-longhorn-engine-with-SPDK.md:19]: Suggest adding a diagram describing the data path from an application to the underlying disk of the current Longhorn arch and the new design.

Comment (DamiaSan) [reply enhancements/20221213-reimplement-longhorn-engine-with-SPDK.md:19]: Added

File: enhancements/20221213-reimplement-longhorn-engine-with-SPDK.md (hunk 1 above)
Comment (DamiaSan) [reply enhancements/20221213-reimplement-longhorn-engine-with-SPDK.md]: Sorry, I didn't explained that SPDK is controlled by JSON-RCP commands, so those I mention in the document are the commands to control SPDK application.
Yes, I mean TCP/IP RPC.

**Comment by DamiaSan** [comment]
Moved to PR 5115 https://github.com/longhorn/longhorn/pull/5115

**Pull request closed without merging**